payara/Payara
# PAYARA-4176 Test containers tests and infrastructure

## Notes for reviewers

- please disable whitespace changes in diff - generally I'm removing trailing spaces and mixed tab and spaces.

## Description

This is a PR introducing TestContainers-based tests, created based on my work used for years on other projects (one had more than 5000 tests in it, useful for

- testing much more possible user scenarios (ie clusters, start/stop instances, hazelcast migrations, destructive tests, logging tests, env-dependent tests ...)
- creating low cost reproducers
- providing examples, covering use cases
- testing from your favourite editor (ie. can be executed directly from Eclipse)

You have also much better control over the environment, it is also easier to ensure the lifecycle of server instance(s).

This PR is **NOT** useful for testing Payara's docker images directly. They have their own tests using TestContainers. This PR is targeting Payara Server features, with or without applications. All other tests we already have usually test applications deployed on Payara (aka "does this application still work with this version?"), rather than Payara Server directly ("does the Payara Server work?").

This PR was split from https://github.com/payara/Payara/pull/4263

## Important Info

### Limitations

- depends on support of the operating system
- depends on installed and properly configured [docker application](https://www.testcontainers.org/supported_docker_environment/windows/). 

### Testing Environment
Kubuntu 20.10
Docker 19.03.13, API 1.40

#### Linux

* your user must be in the docker group, see https://docs.docker.com/install/linux/linux-postinstall/#configure-where-the-docker-daemon-listens-for-connections :

`usermod -aG docker dmatej`

* to run DockerNodeITest, you have to enable access to the dockerd service.
* TODO in another issue: this is insecure practice, so we need to find a secure solution.

```
mkdir -p /etc/systemd/system/docker.service.d
vim /etc/systemd/system/docker.service.d/override.conf
```

```
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -H tcp://0.0.0.0:2376 -H unix:///var/run/docker.sock
```

```
systemctl daemon-reload
service docker restart
curl --verbose http://127.0.0.1:2376/version

```

#### Windows

* see https://www.testcontainers.org/supported_docker_environment/windows/
* read also previous chapter, I don't know how much is required to run it on Windows too

### Blockers and Dependencies
* several tests are disabled, because they depend on fixes under CUSTCOM-55, the PR will be based on this.

### Structure
- test-containers-tools
  - basic infrastructure
- test-containers-arquillian-tools
  - originally used in tests, but when demand for this increased it was pushed outside
  - still interesting thing, I believe with some additional work it is able to simplify all remote testing with arquillian.
- test-containers-manual-test
  - manual testing
- test-containers-tests-cloud
  - main module aggregating test
  - no need of arquillian to write tests with more standalone modules but still maintained lifecycle
  - author of the test has complete control

## Testing
### New tests
All you can find in new packages.

### Testing

First checkout and build everything without tests - this is because you need also parent poms to be available in your local repository cache. Some tests use several different docker images ie to verify behavior of clusters.

```
git remote add dmatej git@github.com:dmatej/Payara.git
git fetch dmatej 
git checkout dmatej/PAYARA-4176-TestContainers
mvn clean install -PBuildExtras,BuildDockerImages -Dmaven.test.skip=true -TC4 # time: 4:37
```

#### Shortest way to retest changes in Payara
Why: I have changed something in Payara, even wrote a unit test, but I need to verify if it did not break something else
```
mvn clean install -PQuickBuild -Dmaven.test.skip=true -TC4 # time: 1:28
mvn clean install -pl :docker-images -PBuildExtras,BuildDockerImages -amd -Dmaven.test.skip=true # time: 1:29
mvn clean install -Ptest-containers -pl :test-containers -amd -fae; # time: 7:34
```

#### Build and test against an older Payara version
Why: I want to check if results of my current tests are same
```
git checkout payara-server-5.2020.5
mvn clean install -pl :docker-images-basic -PBuildExtras,BuildDockerImages -amd -Dmaven.test.skip=true
git checkout PAYARA-4176-TestContainers
mvn clean install -Ptest-containers -pl :test-containers -amd -fae; # time: 7:34, VerboseLoggingITest should fail
```

#### Build and test against not yet released version
```
git checkout payara/Payara-5.2020.7-Release
mvn clean install -Dmaven.test.skip=true -TC4 -PBuildExtras,BuildDockerImages
mvn clean install -pl :docker-images -PBuildExtras,BuildDockerImages -amd -Dmaven.test.skip=true
git checkout PAYARA-4176-TestContainers
mvn clean install -Ptest-containers -pl :test-containers -amd -fae -Ddocker.payara.version=5.2020.7
```
#### Build and test against Payara4
- does not work for Payara4 any more (but before changes for JDK11 in docker-images it did)

#### Manual testing
- no need to do anything on your environment, just run
  ```
    mvn clean install -Ptest-containers,manual -amd -fae -pl :test-containers-manual-test`
  ```
- you will see links to server's endpoints in the console
- server.log is printed to the console
- see logging-old.properties and logging-new.properties (will come later with CUSTCOM-55 PR)
- the file system is linked to the host, so even if you stop the container, you can still open files and usually even start the domain locally (no guarantee as hosts and ip addresses would change, but it may help sometimes)
- CTRL+C stops the build including all running containers, killing the JVM should do that too.
![image](https://user-images.githubusercontent.com/302904/112728784-34f78580-8f29-11eb-8f09-3db633315647.png)
- it is even possible to run it directly from the Eclipse editor - it can be stopped by the red square on the Console view (and needs around 15 seconds to stop)
![image](https://user-images.githubusercontent.com/302904/112728901-c109ad00-8f29-11eb-9ab7-19650524b6b6.png)

#### Manual testing with really old version

```
mvn clean install -Ptest-containers,manual -pl :docker-containers-for-tests,:test-containers-manual-test -Ddocker.payara.version=5.193
```

### Tests

* DockerNodesITest 
  - example of "docker in docker" 
  - examples for https://docs.payara.fish/community/docs/5.2020.6/documentation/payara-server/docker/docker-instances.html
  - I wrote it when I reviewed the implementation.
  - test for changes in https://github.com/payara/Payara/pull/4260
* CorbaXaTxITest
  - example of testing CORBA XA transactions (CUSTCOM-148 bug was not reproduced)
* JdbcPoolITest
  - test covering https://blog.payara.fish/pooling-for-databases-on-the-same-database-server
  - uses also MySqlContainer
  - combining JUnit5 extensions - in this example with DBUnit
* CertificateRealmITest
  - test covering this PR: https://github.com/payara/Payara/pull/4272
  - covers also HTTPS communication with the remote server
  - workarounded bug, issue created in JIRA
  - also already detected bug we created before the release
* VerboseLoggingITest 
  - reproducer for https://github.com/payara/Payara/pull/4360
  - since that time it detected much more errors, many of them were fixed under CUSTCOM-55, some fixes are already in master. 
* JtaTimeoutLoggingITest 
  - tests for fixes around https://github.com/payara/Payara/pull/4323 and https://github.com/payara/Payara/pull/4325
  - disabled, in 90% fails in all releases because of race conditions fixed by CUSTCOM-55, different PR will come soon.
